lib/marisa/exception.cc

```diff
@@ -0,0 +1,21 @@
+#include "marisa/exception.h"
+
+namespace marisa {
+
+/**
+ * This is defined here for the following reasons.
+ *
+ * 1. Virtual destructors are best not defined in a header to be included in many compilation units.
+ * Defining it here in a single compilation unit prevents multiple vtables from being generated and causing
+ * a compiler to be confused as to which is the true vtable, like with full link time optimization.
```

Comment:
Doesn't the linker just merge the vtables? What goes wrong?

Reply:
From my experience, no not really. Especially not when you use one in a shared library and then you include the same header in your application using that shared library. You get symbol conflicts. AIX is very finicky and has been vocal about name conflicts like this in the past when linking.

When you're using clang and using full link time optimization, I believe that it will collapse the vtables. Unfortunately when you try to cast a pointer to determine if the object is of a specific subclass, it may not use the collapsed table. This caused a lot of problems in our code in the past. So we use the -Wweak-vtables warning to avoid this issue in the future. That warning complains about this specific class, and we don't want to see this warning in our code.